Implement github-actions-controller at minimal (#1)
cybozu-go/github-actions-controller: docs/design.md

Design notes
============

Motivation
----------

To run GitHub Actions self-hosted runners faster and stably by making full use of

> the machine resource in the staging environment.

Reviewer: I don't think "staging environment" would make sense to anyone but us.

Goals
-----

- Deploy and manage GitHub Actions self-hosted runners on multiple bare-metal

>   servers easily with Kubernetes.

Reviewer: Can this product be used with managed Kubernetes?

Author: Not limited to bare-metal k8s, so I'm gonna fix it.

- Enable runners to finish a time-consuming initialization step before jobs are
  assigned in order not to make users wait longer.
- Extend lifetime of runners from outside when jobs are failed, to investigate
  what causes the failure.
- Notify users whether jobs are failed or not via Slack and extend the lifetime

>   from Slack.

Reviewer: Can you have a pluggable architecture that allows us to use notifications other than slack?

Author: Currently No, but this architecture is simple and easy to extend, so that might be possible in the future.

Components
------------

- Runner: `Pod` to run GitHub Actions self-hosted runner on
- Controller manager
  - `RunnerPool` reconciler: A controller for the `RunnerPool` custom resource(CR).
  -  `Pod` Mutating webhook: A mutating webhook to inject a GitHub Actions
     registration token to `Pod` `env`.

>   - Runner sweeper: A periodical task runner to sweep runners's registration

Reviewer: I little confused by the fact that "Runner" is used in multiple means.

Author: Thank you. I'm gonna make the definition of the term `Runner` clearer.

Author: Instead, I removed the word runner for the meaning of `A periodical task runner` completely.

    records via GitHub API.
  - `Pod` sweeper: A periodical task runner to sweep Pods which exceeds the deletion
    time limit.
- Slack agent: Server for notifying user whether jobs are failed or not and
  accepting extending requests from outside of Kubernetes.

Diagram: T.B.A

Architecture
------------

### How self-hosted runners are created and runs jobs

1. The `RunnerPool` reconciler watches `RunnerPool` creation events and creates
  a `Deployment` which own runner `Pod`s. The controller adds some labels to
  `Pod`s to tell other components like webhook that the `Pod`s are managed by
  the controller. After this steps, Other components recognize which `Pod`s
  they should handle with those labels.
1. The mutating webhook creates registration tokens via GitHub API and injects
  them to the `Pod`s' `env` fields.
1. Each runner `Pod` runs the following commands.

>   1. Register itself as a self-hosted runner with the injected token.

Reviewer: List rendering is broken.

>   1. Initialize runner environment by doing the user-defnied process.

Reviewer 2: ```suggestion
  1. Initialize runner environment by doing the user-defined process.
```

  1. Start a long polling process and wait for GitHub Actions to assign a job.
  1. Run an assigned job.
  1. Call the Slack agent to notify users.
    - GitHub API does not seem to provide a way to know which runner ran a

>       succeded or failed job.

Reviewer 2: ```suggestion
      succeeded or failed job.
```

    - This repository provides a simple `job-failed` command, and asks users to
      execute this command when the job is failed. The `if: failure()` syntax
      allows users to run a step only when one of previous steps exit with non-zero
      code.
  1. Annotate the `Pod` manifest for itself with a timestamp when to delete this
    `Pod`.

>     - If the job is succeded or canceled, the `Pod` annotates itself with the

Reviewer 2: ```suggestion
    - If the job is succeeded or canceled, the `Pod` annotates itself with the
```

      current time.
    - If the job is failed, the `Pod` annotate itself with the future time, for
      example 20 min later.
1. The Slack agent notifies the result of the job on a Slack channel.
1. Users can extend the failed runner if they want to by clicking a button on Slack.
1. The Slack agent is running a WebSocket process to watch extending messages.

>   If it receives a message, it annotate the `Pod` manifest with the designated time.

Reviewer: ```suggestion
  If it receives a message, it annotates the `Pod` manifest with the designated time.
```

1. `Pod` sweeper periodically checks if there are `Pod`s annotated with the old
  timestamp, and if any, it deletes `Pod`s.

### Runner's state management

A runner `Pod` has the following state as a GitHub Actions job runner.

> - registerred: `Pod` registerred itself on GitHub Actions.

Reviewer: ```suggestion
- registered: `Pod` registered itself on GitHub Actions.
```

- initialized: `Pod` finished the initialization.
- listening: `Pod` starts listening with Long Polling.
- assigned: `Pod` is assigned a job and starts running it.
- debug: The job has finished with failure and Users can enter `Pod` to debug.

Runner `Pod`s might not need to manage all of these state, but some of thes states
are useful to check the system is properly running or not.
The state "debug" is a must to be watched to keep runner `Pod`s in stock.

> Users can leave many failed runner `Pod`s for investigation, but the number of

Reviewer 2: If all the replicas are occupied for the user's debug, does that mean that the self-hosted runner will not run workflow anymore?

Author: Yes, but autoscaling might help us avoid this situation. It's future work.

> replicas is also limited. Exposing the state as Prometheus metrics alllows users

Reviewer: ```suggestion
replicas is also limited. Exposing the state as Prometheus metrics allows users
```

to aggregate the number of `Pod`s in each state and raise alerts that they are
running out of available runners.

However, as mentioned above, GitHub API does not provide a way to connect the job

> status infomataion with the self-hosted runner.  So, runner `Pod`s have to execute

Reviewer: ```suggestion
status informataion with the self-hosted runner.  So, runner `Pod`s have to execute
```

> status infomataion with the self-hosted runner.  So, runner `Pod`s have to execute

Reviewer: ```suggestion
status information with the self-hosted runner.  So, runner `Pod`s have to execute
```

commands to tell the metrics exporter server that the state has changed.

### Why is webhook needed ?

This is mainly because GitHub Actions registration token expires after 1 hour as written [here](https://docs.github.com/en/rest/reference/actions#create-a-registration-token-for-a-repository).

- If a token is given in the `env` field in a pod template of `Deployment`, `Pod`
  cannot register itself again when recreated.
- If a token is given with a `Secret` resource mounted on the `Pod`, recreated
  `Pod`s can access an updated token.
  However, webhook is simpler because only the webhook is responsible for tokens.
  Both a controller and a periodical runner have to take responsibility to manage
  tokens in this alternative design.
- If a token is given in the `env` field in a pod template of `Deployment` and
  `Pod` restarts by having liveness probe fail, `Pod`s do not need to re-register
  itself to GitHub Actions.
  However, if `Pod`s are recreated in some reasons, it cannot register itself again.

> - Persistent volumes can store the resitration infomation, but stateless workloads

Reviewer: ```suggestion
- Persistent volumes can store the registration information, but stateless workloads
```
